### DISCUSSION TITLE
Integrate Android AOSP putting mobile under control

### DISCUSSION THREAD

**Discussion by blaggacao** [nrdxp/nixdev#182] [Ideas]
Users of `devos`, when starting to build their environment, might realize that they have a weak spot: google still tracks them on their smartphone.

I don't really believe Mobile NixOS can be a viable alternative in the short to mid term, but there is robotnix to build custom android images integrating well established third party software of that space into the nix ecosystem.

See announcement: https://discourse.nixos.org/t/robotnix-build-android-aosp-using-nix/6906

/cc @danielfullmer

**Comment by blaggacao** [comment]
@danielfullmer could you justify inclusion into `devos` (to kick start adoption) under the NGI Zero grant? That would be awesome and `devos` is defenitly the right horse to bet on! I can tell, I'm already all-in on this bet for a while. 😄 

@nrdxp, while very skillful himself, might probably be grateful to have support from the proper upstream experts.

**Comment by tgunnoe** [comment]
Do you mean build android images from this repository, much like building ISOs for the configs? 

How does this android relate to this repository and google tracking users on their smartphone?

With Robotnix, does building get around having to accept the Google SDK license? Last year working with flutter and nix, this was a huge problem.

**Comment by blaggacao** [comment]
> Do you mean build android images from this repository, much like building ISOs for the configs?

Yes, that would have been my first guess, as well.

> How does this android relate to this repository and google tracking users on their smartphone?

I think was safe to assume that NixOS users are inclined towards liken full control over their environment, either for value based reasons (privacy, philosophy) or for practical reasons (harness the power of nix-lang, software security). If that is true for nix users, then `devos` perfectly caters those users as it makes a unified environment based on nix-lang a lot, lot easier.

By the way, I don't see any attachment to "NixOS-**only**" as a particular compelling (hypothetic) value proposition of this repo, so I think any _complementary_ gear that helps control one's environment is well fit. Sure, as for the main OS, NixOS has a special place in our all hearts. 😄 

> With Robotnix, does building get around having to accept the Google SDK license?

As I understand it, there is [microG](https://microg.org/) or even [GrapheneOS](https://grapheneos.org/) or [LineageOS](https://lineageos.org/) to completely eskew google. I have no concrete experience though and only recite from my own interpretation of the grand scheme of things.

**Comment by danielfullmer** [comment]
@blaggacao  I wasn't familiar with `devos` until you mentioned it, but I was wondering what more specifically you might have had in mind for integration?  From what I could see, the configuration and deployments in devos would apply just to NixOS systems.

At least tangentially related: Robotnix has recently added support for Nix flakes, with an included [template](https://github.com/danielfullmer/robotnix/blob/master/template/flake.nix) to make it easy to get started. My personal flake-based config is available [here](https://github.com/danielfullmer/robotnix-personal).  I intend to add some additional NixOS modules to the robotnix flake to make it easy for a NixOS system to run some supporting services for the running Android systems. (e.g. OTA update server, attestation server, etc).

> With Robotnix, does building get around having to accept the Google SDK license? Last year working with flutter and nix, this was a huge problem.

@tgunnoe It's possible to build the Android SDK (not including the Google extensions) with robotnix, which I don't believe requires accepting the SDK license. See also: https://github.com/NixOS/nixpkgs/issues/75603#issuecomment-620176766

**Comment by blaggacao** [comment]
> I was wondering what more specifically you might have had in mind for integration?

As a user of devos, I want to control my entire technological environment through a nix-based descriptive (and reproducible) model. Just as I add a package to a devos profile, I would wish to be able to install an application decoratively on my phone as well.

Since android and linux (specifically NixOS) are a few blocks a part, I would need to live with the fact that I cannot reuse (most of) my NixOS configuration for android, and I would have to content with buidling a parallel configuration for my android devices.

Being able to do so via robotnix through devos, however, allows me to have a _single pane of glass_ into my environment and eventually I might even be able to reuse some (higher level) parts of the configuration. In any case, I would be able to access builders and deployers through the _same set of **familiar** proxy entry points_ (currently the `flk` command).

As developer of Robotnix, I can make the project easily accessible to a growing number of devos users and can participate in shaping this "evolving best practice" from a mobile phone / android point of view.

As some parts of devos users (eg. me), I would not be able to come up with a stable / maintainable solution to plug devos & robotnix together on my end. And if I would, that would be a siloed solution without any community mind share or best practice reverberations across slightly more brilliant minds.
